src/crimson/os/seastore/seastore.cc

```diff
@@ -859,7 +874,36 @@ seastar::future<std::map<uint64_t, uint64_t>> SeaStore::fiemap(
   uint64_t off,
   uint64_t len)
 {
-  return seastar::make_ready_future<std::map<uint64_t, uint64_t>>();
+  LOG_PREFIX(SeaStore::fiemap);
+  DEBUG("oid: {}, off: {}, len: {} ", oid, off, len);
+  return repeat_with_onode<std::map<uint64_t, uint64_t>>(
+    ch,
+    oid,
+    Transaction::src_t::READ,
+    "fiemap_read",
+    op_type_t::READ,
+    [=](auto &t, auto &onode) -> ObjectDataHandler::fiemap_ret {
+    size_t size = onode.get_layout().size;
+    if (off >= size) {
+      INFO("fiemap offset is over onode size!");
+      return seastar::make_ready_future<std::map<uint64_t, uint64_t>>();
+    }
+    size_t adjust_len = (len == 0)?
+      size - off:
+      std::min(size - off, len);
+
+    return ObjectDataHandler(max_object_size).fiemap(
```

Comment:
This won't work.  The ObjectDataHandler instance will go out of scope before fiemap resolves.  Use a do_with.

Reply:
do the other seastore functions such as read, write, truncate have the same problem? all of these functions don't put ObjectDataHandler instance into do_with.

Reply:
Hmm, read doesn't capture this and so is safe.  truncate, clear, and write do capture this, so they need a do_with since the ObjectDataHandler instance needs to survive.

Reply:
Sorry about not noticing that earlier!

Reply:
@athanatos   just want to make sure, do you mean something like this: 
 return seastar::do_with(
      ObjectDataHandler(max_object_size),
      [=, &t, &onode] (auto &objhandler) {
      return objhandler.fiemap(

Reply:
Yep